zebedee-cms/src/main/java/com/github/onsdigital/zebedee/model/publishing/legacycacheapi/LegacyCacheApiPayloadBuilder.java

```diff
@@ -0,0 +1,205 @@
+            try {
+                if (StringUtils.isEmpty(incomingUrl)) {
+                    return incomingUrl;
+                }
+
+                String updatedUri = incomingUrl.trim();
+                if (StringUtils.isEmpty(updatedUri)) {
+                    return "";
+                }
+
+                // Order matters:
+                // 1. If starts with ons return original
+                if (updatedUri.startsWith(ONS_PREFIX)) {
+                    return updatedUri;
+                }
+
+                // 2. If starts with /visualisation/ then /visualisation/xxx/xxxx/ ---> /visualisation/xxx (keep 1 segment) - return here
+                if (updatedUri.startsWith(VISUALISATIONS_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, VISUALISATIONS_SEGMENT, 1);
+                }
+
+                // 3. If resource then e.g. /embed?uri='/something' ---> /something:
+                //      /chartconfig
+                //      /chartimage
+                //      /embed
+                //      /chart
+                //      /resource
+                //      /generator
+                //      /file
+                //      /export
+                for (String keyword : RESOURCE_TYPE_SEGMENTS) {
+                    if (updatedUri.startsWith(keyword)) {
+                        updatedUri = URIUtils.getQueryParameterFromURL(updatedUri, "uri");
+                        break;
+                    }
+                }
+
+                // 4. If bulletin or article then /xxxx/bulletin/xxxx1/xxxx2/xxxx3 --> /bulletin/xxxx1/xxxx2  (keep 2 segment) & RETURN modified
+                if (updatedUri.contains(ARTICLE_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, ARTICLE_SEGMENT, 2);
+                }
+                if (updatedUri.contains(BULLETIN_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, BULLETIN_SEGMENT, 2);
+                }
+                // 5. If qmis or adhoc then /xxxx/qmis/xxx1/xxx2 --> /qmis/xxx1 & RETURNS modified
+                if (updatedUri.contains(ADHOC_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, ADHOC_SEGMENT, 1);
+                }
+                if (updatedUri.contains(QMIS_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, QMIS_SEGMENT, 1);
+                }
+
+                // 6. For the below if (a) is true, skip (b):
+                //      (a) If path ends with /data or /data/xxx then /xxx1/data/ --> /xxx1
+                //      (b) If path ends with filename + extension remove it /xxx/filename.extension --> /xxx
+                if (URIUtils.getLastSegment(updatedUri) != null) {
+                    if (URIUtils.getLastSegment(updatedUri).matches("^" + DATA_SEGMENT_WITHOUT_SLASH + "/?$")) {
```

Comment:
You shouldn't need regular expressions for this. All you need to do is to check if the last segment is "data". If you remove the trailing slash after step 3 (extracting the URI), then you'll know that from that point on the URI won't have a trailing slash, which will make your code easier. It's also a requirement to remove the trailing slash anyway.